From cb99aaa13a013787532eec4a73588b50dd65b132 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Thu, 18 Aug 2022 14:15:10 +0300
Subject: [PATCH 080/151] clk: renesas: r8a779f0: improve control domain stub

Read register values to find out exact parents and dividers for some
clocks. But still don't modify anything.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 drivers/clk/renesas/r8a779f0-cpg-mssr.c | 97 ++++++++++++++++++++++---
 drivers/clk/renesas/renesas-cpg-mssr.c  |  7 ++
 drivers/clk/renesas/renesas-cpg-mssr.h  |  4 +
 3 files changed, 99 insertions(+), 9 deletions(-)

diff --git a/drivers/clk/renesas/r8a779f0-cpg-mssr.c b/drivers/clk/renesas/r8a779f0-cpg-mssr.c
index 0c3b0ead7279..447d8369c992 100644
--- a/drivers/clk/renesas/r8a779f0-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a779f0-cpg-mssr.c
@@ -29,6 +29,9 @@ enum r8a779f0_clk_types {
 	CLK_TYPE_CD_FIRST = CLK_TYPE_GEN4_SOC_BASE,
 	CLK_TYPE_CD_MOSC = CLK_TYPE_CD_FIRST,
 	CLK_TYPE_CD_PLL,
+	CLK_TYPE_CD_PLLO,
+	CLK_TYPE_CD_SYS,
+	CLK_TYPE_CD_UHSB,
 };
 
 enum clk_ids {
@@ -143,18 +146,30 @@ static const struct cpg_core_clk r8a779f0_core_clks[] __initconst = {
 	DEF_GEN4_MDSEL("r",	R8A779F0_CLK_R, 29, CLK_EXTALR, 1, CLK_OCO, 1),
 
 	/* Control domain internal clocks */
+
+	/* rate of CLK_MOSC is fixed, depends on MD 14:13 */
 	DEF_TYPE("cd_mosc",	CD_CLK_MOSC,	CLK_TYPE_CD_MOSC),
+
+	/* rates of CLK_HSIOSC, CLK_LSIOSC, CLK_HVIOSC are fixed */
 	DEF_RATE("cd_hsiosc",	CD_CLK_HSIOSC,	200 * 1000 * 1000),
 	DEF_RATE("cd_lsiosc",	CD_CLK_LSIOSC,	240 * 1000),
 	DEF_RATE("cd_hviosc",	CD_CLK_HVIOSC,	4 * 1000 * 1000),
+
+	/* rate of CLK_PLL is fixed, depends om MD 40:39 */
 	DEF_TYPE("cd_pll",	CD_CLK_PLL,	CLK_TYPE_CD_PLL),
-	/* With the default setting, CLK_PLLO is not divided */
-	DEF_FIXED("cd_pllo",	CD_CLK_PLLO,	CD_CLK_PLL, 1, 1),
-	/* CLK_IOSC sourced from CLK_HSIOSC unless that is stopped */
+
+	/* CLK_PLLO is divided from CLK_PLL */
+	DEF_TYPE("cd_pllo",	CD_CLK_PLLO,	CLK_TYPE_CD_PLLO),
+
+	/* CLK_IOSC sourced from CLK_HSIOSC unless that is stopped, which
+	 * happens only in complex standby modes */
 	DEF_FIXED("cd_iosc",	CD_CLK_IOSC,	CD_CLK_HSIOSC, 1, 1),
-	/* With the default setting, CLK_SYS is sourced from CLK_IOSC */
-	DEF_FIXED("cd_sys",	CD_CLK_SYS,	CD_CLK_IOSC, 1, 1),
-	DEF_RATE("cd_uhsb",	CD_CLK_UHSB,	160 * 1000 * 1000),
+
+	/* CLK_SYS sourced from either CLK_PLLO or CLK_IOSC */
+	DEF_TYPE("cd_sys",	CD_CLK_SYS,	CLK_TYPE_CD_SYS),
+
+	/* CLK_UHSB / CLK_HSB / CLK_LSB defined per Table 13.6 */
+	DEF_TYPE("cd_uhsb",	CD_CLK_UHSB,	CLK_TYPE_CD_UHSB),
 	DEF_FIXED("cd_hsb",	CD_CLK_HSB,	CD_CLK_UHSB, 2, 1),
 	DEF_FIXED("cd_lsb",	CD_CLK_LSB,	CD_CLK_UHSB, 4, 1),
 
@@ -295,30 +310,52 @@ static int __init r8a779f0_cpg_mssr_init(struct device *dev)
 	return rcar_gen4_cpg_init(cpg_pll_config, CLK_EXTALR, cpg_mode);
 }
 
+#define CLKD_PLLC_OFFSET		0x120
+#define PLLO_DIV_SHIFT			0
+#define PLLO_DIV_WIDTH			3
+#define PLLO_PARENT			"cd_pll"
+static const struct clk_div_table pllo_div_table[] = {
+	{ .val = 1, .div = 1 },
+	{ .val = 2, .div = 2 },
+	{}
+};
+
+#define CLKC_CPUS_OFFSET		0x108
+#define SYS_PARENT_SHIFT		0
+#define SYS_PARENT_WIDTH		1
+static const char * const sys_parents[] = { "cd_pllo", "cd_iosc" };
+
 static struct clk * __init r8a779f0_cpg_clk_register(struct device *dev,
 	const struct cpg_core_clk *core, const struct cpg_mssr_info *info,
 	struct clk **clks, void __iomem *base,
 	struct raw_notifier_head *notifiers)
 {
 	u64 mode_pins;
+	int md_bits;
 	int error;
 	unsigned long rate;
+	int div;
 
 	switch (core->type) {
 	case CLK_TYPE_CD_MOSC:
 		error = rcar_rst_read_mode_pins_64(&mode_pins);
 		if (error)
 			return ERR_PTR(error);
-		switch ((mode_pins >> 13) & 3) {
+		md_bits = (mode_pins >> 13) & 3;
+		switch (md_bits) {
 		case 0:
 			rate = 16 * 1000 * 1000;
 			break;
 		case 1:
 			rate = 20 * 1000 * 1000;
 			break;
-		default:
+		case 3:
 			rate = 40 * 1000 * 1000;
 			break;
+		default:
+			dev_err(dev, "Prohibited setting: md 14:13 is %d\n",
+					md_bits);
+			return ERR_PTR(-EINVAL);
 		}
 		return clk_register_fixed_rate(NULL, core->name, NULL, 0, rate);
 
@@ -326,7 +363,8 @@ static struct clk * __init r8a779f0_cpg_clk_register(struct device *dev,
 		error = rcar_rst_read_mode_pins_64(&mode_pins);
 		if (error)
 			return ERR_PTR(error);
-		switch ((mode_pins >> 39) & 3) {
+		md_bits = (mode_pins >> 39) & 3;
+		switch (md_bits) {
 		case 3:
 			rate = 800 * 1000 * 1000;
 			break;
@@ -339,6 +377,47 @@ static struct clk * __init r8a779f0_cpg_clk_register(struct device *dev,
 		}
 		return clk_register_fixed_rate(NULL, core->name, NULL, 0, rate);
 
+	case CLK_TYPE_CD_PLLO:
+		return clk_register_divider_table(NULL, core->name,
+				PLLO_PARENT, 0,
+				cpg_mssr_cd_base(dev) + CLKD_PLLC_OFFSET,
+				PLLO_DIV_SHIFT, PLLO_DIV_WIDTH,
+				CLK_DIVIDER_READ_ONLY,
+				pllo_div_table, NULL);
+
+	case CLK_TYPE_CD_SYS:
+		return clk_register_mux(NULL, core->name,
+				sys_parents, ARRAY_SIZE(sys_parents), 0,
+				cpg_mssr_cd_base(dev) + CLKC_CPUS_OFFSET,
+				SYS_PARENT_SHIFT, SYS_PARENT_WIDTH,
+				CLK_MUX_READ_ONLY, NULL);
+
+	case CLK_TYPE_CD_UHSB:
+		if (ioread32(cpg_mssr_cd_base(dev) + CLKC_CPUS_OFFSET) & 1) {
+			/* CLK_SYS sourced from CLK_IOSC - divider depends
+			 * on MD 40:39 */
+			error = rcar_rst_read_mode_pins_64(&mode_pins);
+			if (error)
+				return ERR_PTR(error);
+			md_bits = (mode_pins >> 39) & 3;
+			switch (md_bits) {
+			case 3:
+				div = 5;	/* 200 -> 40 */
+				break;
+			case 2:
+				div = 4;	/* 200 -> 50 */
+				break;
+			default:
+				div = 3;	/* 200 -> 66.67 */
+				break;
+			}
+		} else {
+			/* CLK_SYS sourced from CLK_POOL - divider is 5 */
+			div = 5;
+		}
+		return clk_register_fixed_factor(NULL, core->name,
+					"cd_sys", 0, 1, div);
+
 	default:
 		return rcar_gen4_cpg_clk_register(dev, core, info, clks, base,
 				notifiers);
diff --git a/drivers/clk/renesas/renesas-cpg-mssr.c b/drivers/clk/renesas/renesas-cpg-mssr.c
index 2725a132a5cb..2bb681148059 100644
--- a/drivers/clk/renesas/renesas-cpg-mssr.c
+++ b/drivers/clk/renesas/renesas-cpg-mssr.c
@@ -1346,5 +1346,12 @@ void __init mssr_mod_reparent(struct mssr_mod_clk *mod_clks,
 		}
 }
 
+void __iomem *cpg_mssr_cd_base(struct device *dev)
+{
+	struct cpg_mssr_priv *priv = dev_get_drvdata(dev);
+
+	return priv->cd_base;
+}
+
 MODULE_DESCRIPTION("Renesas CPG/MSSR Driver");
 MODULE_LICENSE("GPL v2");
diff --git a/drivers/clk/renesas/renesas-cpg-mssr.h b/drivers/clk/renesas/renesas-cpg-mssr.h
index 15fe9c2e526e..e6af770eba0a 100644
--- a/drivers/clk/renesas/renesas-cpg-mssr.h
+++ b/drivers/clk/renesas/renesas-cpg-mssr.h
@@ -214,4 +214,8 @@ extern void mssr_mod_reparent(struct mssr_mod_clk *mod_clks,
 			      unsigned int num_mod_clks,
 			      const struct mssr_mod_reparent *clks,
 			      unsigned int n);
+
+/* HACK: access to control domain register base */
+void __iomem *cpg_mssr_cd_base(struct device *dev);
+
 #endif
-- 
2.17.1

